File: Diagrammatize.Design.Pattern/code/cxx/Facade/Sample/pagemaker/HtmlWriter.cpp
#include "HtmlWriter.hpp"

HtmlWriter::HtmlWriter(std::ostream& writer_): writer(writer_) {
}

void HtmlWriter::title(const std::string& title) {
    writer << "<html>";
    writer << "<head>";
    writer << "<title>" << title << "</title>";
    writer << "</head>";
    writer << "<body>\n";
    writer << "<h1>" << title << "</h1>\n";
}

void HtmlWriter::paragraph(const std::string& msg) {
    writer << "<p>" << msg << "</p>\n";
}

void HtmlWriter::link(const std::string& href, const std::string& caption) {
    paragraph("<a href=\"" + href + "\">" + caption + "</a>");
}

void HtmlWriter::mailto(const std::string& mailaddr, const std::string& username) {
    link("mailto:" + mailaddr, username);
}

void HtmlWriter::close() {
    writer << "</body>";
    writer << "</html>\n";
}

File: Diagrammatize.Design.Patterns/code/python/Facade/Sample/pagemaker/HtmlWriter.py

class HtmlWriter:
    def __init__(self, writer):
        self.writer = writer

    def title(self, title):
        self.writer.write("<html>")
        self.writer.write("<head>")
        self.writer.write("<title>" + title + "</title>")
        self.writer.write("</head>")
        self.writer.write("<body>\n")
        self.writer.write("<h1>" + title + "</h1>\n")

    def paragraph(self, msg):
        self.writer.write("<p>" + msg + "</p>\n")

    def link(self, href, caption):
        self.paragraph("<a href=\"" + href + "\">" + caption + "</a>");

    def mailto(self, mailaddr, username):
        self.link("mailto:" + mailaddr, username)

    def close(self):
        self.writer.write("</body>")
        self.writer.write("</html>\n")
        self.writer.close()
